velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -16,37 +16,62 @@
 #include "velox/functions/lib/Re2Functions.h"
 
 #include "velox/common/base/Exceptions.h"
+#include "velox/expression/VectorWriters.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-// Blocks patterns that contain character class union, intersection, or
-// difference because these are not understood by RE2 and will be parsed as a
-// different pattern than in java.util.regex.
-void ensureRegexIsCompatible(
-    const char* functionName,
-    const VectorPtr& patternVector) {
-  if (!patternVector ||
-      patternVector->encoding() != VectorEncoding::Simple::CONSTANT) {
-    VELOX_USER_FAIL("{} requires a constant pattern.", functionName);
+using ::re2::RE2;
+
+template <typename T>
+re2::StringPiece toStringPiece(const T& s) {
+  return re2::StringPiece(s.data(), s.size());
+}
+
+// If v is a non-null constant vector, returns the constant value. Otherwise
+// returns nullopt.
+template <typename T>
+std::optional<T> getIfConstant(const BaseVector& v) {
+  if (v.encoding() == VectorEncoding::Simple::CONSTANT &&
+      v.isNullAt(0) == false) {
+    return v.as<ConstantVector<T>>()->valueAt(0);
   }
-  if (patternVector->isNullAt(0)) {
-    return;
+  return std::nullopt;
+}
+
+void checkForBadPattern(const RE2& re) {
+  if (UNLIKELY(!re.ok())) {
+    VELOX_USER_FAIL("invalid regular expression:{}", re.error());
   }
-  const StringView pattern =
-      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+}
+
+/**
+ * Validates the provided regex pattern to ensure its compatibility with the
+ * system. The function checks if the pattern uses features like character class
+ * union, intersection, or difference which are not supported.
```

Comment:
Maybe mention these restrictions in the function documentation.

Reply:
These restrictions and others are mentioned in https://github.com/google/re2/wiki/Syntax link in the Spark regexp documentation